### PR TITLE
Allowed arbitrary CMIP6 fx variables in special preprocessors and added a catch on project not found in config-developer (invalid project)

### LINKED ISSUE #431: Use of areacella and areacello in area_statistics not supported for CMIP6
Using the following preprocessor

```yaml
  spatial_mean_sea:
    mask_landsea:
      mask_out: land
    area_statistics:
      operator: mean
      fx_files: ['areacello']
```

fails at the moment:

```
2020-01-15 13:48:55,511 UTC [22398] ERROR   Program terminated abnormally, see stack trace below for more information
Traceback (most recent call last):
  File "ESMValCore/esmvalcore/_main.py", line 229, in run
    conf = main(args)
  File "ESMValCore/esmvalcore/_main.py", line 157, in main
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "ESMValCore/esmvalcore/_main.py", line 203, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "ESMValCore/esmvalcore/_recipe.py", line 73, in read_recipe_file
    recipe_file=filename)
  File "ESMValCore/esmvalcore/_recipe.py", line 909, in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
  File "ESMValCore/esmvalcore/_recipe.py", line 1255, in initialize_tasks
    task_name=task_name,
  File "ESMValCore/esmvalcore/_recipe.py", line 880, in _get_preprocessor_task
    name=task_name,
  File "ESMValCore/esmvalcore/_recipe.py", line 738, in _get_single_preprocessor_task
    config_user=config_user,
  File "ESMValCore/esmvalcore/_recipe.py", line 689, in _get_preprocessor_products
    config_user=config_user)
  File "ESMValCore/esmvalcore/_recipe.py", line 478, in _update_fx_settings
    for fxvar in var['fx_files']}
  File "ESMValCore/esmvalcore/_recipe.py", line 478, in <dictcomp>
    for fxvar in var['fx_files']}
  File "ESMValCore/esmvalcore/_recipe.py", line 408, in _get_correct_fx_file
    fx_files = _get_input_files(fx_var, config_user)
UnboundLocalError: local variable 'fx_var' referenced before assignment
```

because certain fx variables are hardcoded in `_recipe.py` for CMIP6. I will take care of that.

### PR DESCRIPTION
**Pull request by schlunma**
**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #431.

### CODE REVIEW

**Comment by schlunma** [comment]
I think we can tackle the #429 in a more general way here. Can you test if #405 is fixed in this branch @ledm? I have to add tests for this.

**Comment by schlunma** [comment]
We only raise an error if the fx var is not in the CMOR tables, if only the files are not available no error is raised.

**Comment by valeriupredoi** [comment]
Ow true, my bad, then I'd move the check up where the cmor variable is
built, use a try/except with a debug logger mssg, and let the code go
through even if the cmor variable is not built - if no cmor variable found
return an empty fx_files list so NE masks can still be used

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

**Comment by valeriupredoi** [comment]
Actually, no need for try/except block, just remove the searched_mips and
its use and referencing

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

**Comment by schlunma** [comment]
But why would we need/allow an fx variable name that is not in the CMOR tables? E.g. if you have a typo in your fx name, then it is really hard to figure out what is failing.

**Comment by valeriupredoi** [comment]
Also I believe @ledm case of volcello found in Omon rather than Ofx will
not work now since fx_var can not be constructed with Omon so the file
search will be done with Ofx instead, and return empty

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

**Comment by valeriupredoi** [comment]
Yes but fx_variable will be built with Ofx (and will be None with Omon) and
the data finder will look for data with mip==Ofx and find nada

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

**Comment by schlunma** [comment]
`volcello` is also in the table `Omon`, so `fx_variable` builds just find :wink:

**Comment by valeriupredoi** [comment]
Blast! these fx vars are getting on my nerves :)) well in that case you
better pray Omon comes last so fx_files gets built with that (and it does,
coz you appended it last as the var's mip, but I would be a woosh more
careful about that)

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

**Comment by valeriupredoi** [comment]
Ah dont mind that, just saw the break. Go on then, all good from yapping V
:grin:

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

**Comment by ledm** [comment]
Does this fix address the other problem raised in #405: Some parts of the preprocessor chain need to be applied to the fx file as well as the diagnostic variable data.

**Comment by valeriupredoi** [comment]
Nope, that's a completely dish of curry. I'll think about that today.

Also, apologies for my previous email comments introducing my lengthy email signature :beer:

**Comment by valeriupredoi** [comment]
BTW if I didn't say it - really nice approach @schlunma :beer:

**Comment by schlunma** [comment]
> BTW if I didn't say it - really nice approach @schlunma beer

Thanks :grin:

I was a bit scared of expanding this to all projects, but actually this should be no problem.

One thing I mentioned: Do we check if there's a CMOR table for the project somewhere before this code? If not, then `cmor_table = CMOR_TABLES[var_project]` returns `None` afaik and we get a really nasty `AttributeError` in line 398.

**Comment by valeriupredoi** [comment]
`variable` (or `var` as in our function) is already initialized and `_add_cmor_info` has been run on it so `CMOR_TABLES[var_project]` should always return the correct cmor table, man :beer:

**Comment by valeriupredoi** [comment]
here's a test running with a fictitious `CMIP7` as `variable["project"]`:
```
esmvalcore/_recipe.py:757: in _get_single_preprocessor_task
    config_user=config_user,
esmvalcore/_data_finder.py:270: in get_output_file
    cfg = get_project_config(variable['project'])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

project = 'CMIP7'

    def get_project_config(project):
        """Get developer-configuration for project."""
        logger.debug("Retrieving %s configuration", project)
>       return CFG[project]
E       KeyError: 'CMIP7'

esmvalcore/_config.py:171: KeyError
```

**Comment by schlunma** [comment]
Nice, so not our business :smile:

**Comment by valeriupredoi** [comment]
of course it is, seaman Beaumont! (ref to Hunt for Red October)
Here's the catch and controlled fail :+1: :beer:

**Comment by valeriupredoi** [comment]
BTW the last test I added is a message for future us when we will run ESMValTool on CMIP7 data :rofl:

File: esmvalcore/_config.py [hunk 1]
```diff
@@ -168,7 +168,11 @@ def configure_logging(cfg_file=None, output=None, console_log_level=None):
 def get_project_config(project):
     """Get developer-configuration for project."""
     logger.debug("Retrieving %s configuration", project)
-    return CFG[project]
+    if project in CFG:
+        return CFG[project]
+    else:
+        raise ValueError("Project '{}' not in config-developer".format(
```
Comment (mattiarighi) [inline esmvalcore/_config.py]:
```suggestion
        raise ValueError("Project '{}' not in config-developer.yml".format(
```

**Pull request merged**